File: src/factories/internal/metadata/emend_metadata_atomics.ts
import { Metadata } from "../../../schemas/metadata/Metadata";
import { MetadataAtomic } from "../../../schemas/metadata/MetadataAtomic";

import { ArrayUtil } from "../../../utils/ArrayUtil";

export const emend_metadata_atomics = (meta: Metadata) => {
  // ATOMICS
  for (const a of meta.atomics) {
    if (is_not_pure(a)) continue;
    const index: number = meta.constants.findIndex((c) => c.type === a.type);
    if (index !== -1) meta.constants.splice(index, 1);
  }

  // BOOLEAN
  {
    const index: number = meta.constants.findIndex((c) => c.type === "boolean");
    if (index !== -1 && meta.constants[index]!.values.length === 2) {
      const temp = meta.constants.splice(index, 1)[0]!;
      ArrayUtil.take(
        meta.atomics,
        (a) => a.type === "boolean",
        () =>
          MetadataAtomic.create({
            type: "boolean" as const,
            tags: temp.values[0]!.tags ?? [],
          }),
      );
    }
  }

  // TEMPLATE
  if (meta.templates.length) {
    const atomic: MetadataAtomic | undefined = meta.atomics.find(
      (a) => a.type === "string",
    );
    if (atomic !== undefined && false === is_not_pure(atomic))
      meta.templates.splice(0, meta.templates.length);
  }
};

const is_not_pure = (atomic: MetadataAtomic): boolean =>
  atomic.tags.length !== 0 &&
  atomic.tags.every(
    (row) => row.length !== 0 && row.every((c) => !!c.validate?.length),
  );

File: packages/typia/native/core/factories/internal/metadata/emend_metadata_atomics.go
package metadata

import schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

func Emend_metadata_atomics(meta *schemametadata.MetadataSchema) {
	for _, atomic := range meta.Atomics {
		if is_not_pure(atomic) {
			continue
		}
		index := -1
		for i, constant := range meta.Constants {
			if constant.Type == atomic.Type {
				index = i
				break
			}
		}
		if index != -1 {
			meta.Constants = append(meta.Constants[:index], meta.Constants[index+1:]...)
		}
	}

	index := -1
	for i, constant := range meta.Constants {
		if constant.Type == "boolean" {
			index = i
			break
		}
	}
	if index != -1 && len(meta.Constants[index].Values) == 2 {
		temp := meta.Constants[index]
		meta.Constants = append(meta.Constants[:index], meta.Constants[index+1:]...)
		found := false
		for _, atomic := range meta.Atomics {
			if atomic.Type == "boolean" {
				found = true
				break
			}
		}
		if found == false {
			tags := [][]schemametadata.IMetadataTypeTag{}
			if len(temp.Values) != 0 && temp.Values[0] != nil {
				tags = temp.Values[0].Tags
			}
			meta.Atomics = append(meta.Atomics, schemametadata.MetadataAtomic_create(schemametadata.MetadataAtomic{
				Type: "boolean",
				Tags: tags,
			}))
		}
	}

	if len(meta.Templates) != 0 {
		var atomic *schemametadata.MetadataAtomic
		for _, candidate := range meta.Atomics {
			if candidate.Type == "string" {
				atomic = candidate
				break
			}
		}
		if atomic != nil && is_not_pure(atomic) == false {
			meta.Templates = meta.Templates[:0]
		}
	}
}

func is_not_pure(atomic *schemametadata.MetadataAtomic) bool {
	if len(atomic.Tags) == 0 {
		return false
	}
	for _, row := range atomic.Tags {
		if len(row) == 0 {
			return false
		}
		for _, tag := range row {
			if tag.Validate == "" {
				return false
			}
		}
	}
	return true
}
